sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1910,486 @@ def transform(self, X):
+        # find index for lower and higher bounds
+        lower_bounds_idx = (X_col - BOUNDS_THRESHOLD <
+                            lower_bound_x)
+        upper_bounds_idx = (X_col + BOUNDS_THRESHOLD >
+                            upper_bound_x)
+
+        references = np.linspace(0, 1, self.n_quantiles, endpoint=True)
+        if not inverse:
+            X_col = np.interp(X_col, self.quantiles_[:, feature_idx],
+                              references)
+        else:
+            X_col = np.interp(X_col, references,
+                              self.quantiles_[:, feature_idx])
+
+        X_col[upper_bounds_idx] = upper_bound_y
+        X_col[lower_bounds_idx] = lower_bound_y
+        # for forward transform, match the output PDF
+        if not inverse:
+            X_col = output_distribution.ppf(X_col)
+            # find the value to clip the data to avoid mapping to
+            # infinity. Clip such that the inverse transform will be
+            # consistent
+            clip_min = output_distribution.ppf(BOUNDS_THRESHOLD -
+                                               np.spacing(1))
+            clip_max = output_distribution.ppf(1 - (BOUNDS_THRESHOLD -
+                                                    np.spacing(1)))
+            X_col = np.clip(X_col, clip_min, clip_max)
+
+        return X_col
+
+    def _dense_transform(self, X, inverse=False):
+        """Forward and inverse transform for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+
+        inverse : bool, optional (default=False)
+            If False, apply forward transform. If True, apply
+            inverse transform.
+
+        Returns
+        -------
+      X : ndarray, shape (n_samples, n_features)
+            Projected data        """
+
+        for feature_idx in range(X.shape[1]):
+            X[:, feature_idx] = self._transform_col(X[:, feature_idx],
+                                                    feature_idx, inverse)
+
+        return X
+
+    def _sparse_transform(self, X, inverse=False):
+        """Forward and inverse transform for sparse matrices.
+
+        Parameters
+        ----------
+        X : sparse matrix CSC, shape (n_samples, n_features)
+            The data used to scale along the features axis. The sparse matrix
+            needs to be semi-positive.
+
+        inverse : bool, optional (default=False)
+            If False, apply forward transform. If True, apply
+            inverse transform.
+
+        Returns
+        -------
+        X : sparse matrix CSC, shape (n_samples, n_features)
+            Projected data.
+        """
+
+        for feature_idx in range(X.shape[1]):
+            column_slice = slice(X.indptr[feature_idx],
+                                 X.indptr[feature_idx + 1])
+            X.data[column_slice] = self._transform_col(X.data[column_slice],
```

Comment:
even in the `ignore_implicit_zeros=False` case, we're not transforming implicit zeros to their quantile. This means that `transform(csr_matrix(X)).A` and `transform(csr_matrix(X).A).A` will not be identical (`.A` means `.toarray()`). At a minimum, this behaviour needs to be more prominently documented. One option is to optionally offset the output distribution so that `transform(0) == 0`. Thus we might be better off making sparse matrices raise an error by default. If we want to handle them, we can provide an option, e.g. `sparse_handling='error'/'explicit_values'/'offset'`.